libcnb-cargo/tests/test.rs

```diff
@@ -1,307 +1,330 @@
-    target_triple: String,
-}
+    assert_eq!(
+        &read_buildpack_data(packaged_buildpack_dir)
+            .unwrap()
+            .buildpack_descriptor
+            .buildpack()
+            .id,
+        buildpack_id
+    );
 
-struct TestOutput {
-    stdout: String,
-    stderr: String,
-    code: Option<i32>,
+    assert_eq!(
+        read_buildpackage_data(packaged_buildpack_dir)
+            .unwrap()
+            .unwrap()
+            .buildpackage_descriptor
+            .dependencies,
+        expected_buildpackage_dependencies
+    );
 }
 
-impl BuildpackPackagingTest {
-    fn new(fixture_name: &str, target_triple: &str) -> Self {
-        let source_directory = env::current_dir()
-            .unwrap()
-            .join("fixtures")
-            .join(fixture_name);
-        let target_directory = tempdir().unwrap();
-        let copy_options = CopyOptions::new();
-        copy(source_directory, &target_directory, &copy_options).unwrap();
-        BuildpackPackagingTest {
-            fixture_name: fixture_name.to_string(),
-            temp_dir: target_directory,
-            release_build: true,
-            target_triple: String::from(target_triple),
-        }
-    }
+fn create_packaged_buildpack_dir_resolver(
+    cargo_target_dir: &Path,
+    release: bool,
+    target_triple: &str,
+) -> impl Fn(&BuildpackId) -> PathBuf {
+    let cargo_target_dir = PathBuf::from(cargo_target_dir);
+    let target_triple = target_triple.to_string();
 
-    fn dir(&self) -> PathBuf {
-        self.temp_dir
-            .path()
-            .canonicalize()
-            .unwrap()
-            .join(&self.fixture_name)
+    move |buildpack_id| {
+        cargo_target_dir
+            .join("buildpack")
+            .join(&target_triple)
+            .join(if release { "release" } else { "debug" })
+            .join(buildpack_id.as_str().replace('/', "_"))
     }
+}
 
-    fn target_dir_name(&self, buildpack_id: BuildpackId) -> String {
-        self.target_dir(buildpack_id)
-            .canonicalize()
-            .unwrap()
-            .to_string_lossy()
-            .to_string()
-    }
+fn copy_fixture_to_temp_dir(name: &str) -> Result<TempDir, std::io::Error> {
+    let fixture_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+        .join("fixtures")
+        .join(name);
 
-    fn target_dir(&self, buildpack_id: BuildpackId) -> PathBuf {
-        get_buildpack_target_dir(
-            &buildpack_id,
-            &self.dir().join("target"),
-            self.release_build,
-            &self.target_triple,
-        )
-    }
+    env::temp_dir()
+        // libcnb-cargo itself will canonicalize paths too.
+        // If we don't, the string representations of the path won't match. Doing this here will
+        // simplify code at the call-sites since they don't have to canonicalize themselves.
+        // Ref: https://github.com/rust-lang/rust/issues/99608
+        .canonicalize()
+        .and_then(tempdir_in)
```

Comment:
So that we can canonicalize the path. See the comment and ref above, probably needs more explanation if that doesn't make it clear enough. We can't canonicalise later because we work with `TempDir`, not `PathBuf`. By explicitly canonicalising the `temp_dir()` result and telling `tempfile` to create the temporary directory in there we can. :)

This is mostly a macOS issue (or any other fs where the `env::temp_dir` points to a symlink.

Reply:
Ah I'd read the comment, but it was above the `canonicalize()` line so it wasn't clear it was referring to the approach in general. 

I think maybe move the comment above `env::temp_dir()` and mention something like: (a) we need to canonialize paths because of OSes where the tmpdir can be a symlink, (b) that we can't use `tempfile::tempdir` if we need to canonicalize?

Reply:
Attempted fix: [308f853](https://github.com/heroku/libcnb.rs/pull/637/commits/308f853fdcbe1aeba94a30f0815946511ac40cbb)